### PR TITLE
categorical binning backend

### PR DESCRIPTION
**Pull request by gbordyugov**
After having read about tricky heuristics to solve the NP-hard problem of binning, I decided to go with a simplest greedy algorithm: sort the items in the decreasing order of their weight and then assign them to the current lightest bin.

This is just a backend of the binning, i.e. a function that return a list of pairs `(weight, list_of_items)`. Need to wrap it into classes, I guess.

### CODE REVIEW

File: expan/core/categorical_binning.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+        if not item in weights:
+            weights[item] = 1
+        else:
+            weights[item] += 1
+
+    # we need items sorted in decreasing order
+    pairs = sorted([(weight, [item]) for (item, weight) in weights.items()], reverse=True)
+
+    # take max(n_bins, len(pairs)) haviest pairs as initial bins
+    bins = [(weight, labels) for (_, (weight, labels)) in zip(range(n_bins), pairs)]
```
Comment (igusher) [inline expan/core/categorical_binning.py]: is it the same as pairs[:n_bins] ?

Comment (gbordyugov) [reply expan/core/categorical_binning.py]: not quite.

`pairs` might happen to be have fewer than `n_bins` elements (`zip` would terminate at the shortest of two)

**Comment by igusher** [comment]
:+1:

**Comment by daryadedik** [comment]
👍

**Review by daryadedik — APPROVED** [review]

**Comment by shansfolder** [comment]
integrated into PR #155

**Pull request closed without merging**